File: app01_java_nft/backend/src/main/java/com/devpowers/nftcerts/identity/MockKycVerificationService.java
package com.devpowers.nftcerts.identity;

import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.UUID;

/**
 * Mock KYC provider integration: always approves verification. Stands in for a real DID/KYC
 * provider (e.g. Persona, Civic) that would be wired in a production deployment.
 */
@Service
public class MockKycVerificationService implements KycVerificationService {

    private final ArtistIdentityRepository artistIdentityRepository;

    public MockKycVerificationService(ArtistIdentityRepository artistIdentityRepository) {
        this.artistIdentityRepository = artistIdentityRepository;
    }

    @Override
    public ArtistIdentity verify(String walletAddress, String did, String email) {
        ArtistIdentity identity = artistIdentityRepository.findByWalletAddress(walletAddress)
                .orElseGet(() -> new ArtistIdentity(UUID.randomUUID(), walletAddress, did, email, false, null));

        identity.setDid(did);
        identity.setEmail(email);
        identity.setVerified(true);
        identity.setVerifiedAt(Instant.now());

        return artistIdentityRepository.save(identity);
    }

    @Override
    public boolean isVerified(String walletAddress) {
        return artistIdentityRepository.findByWalletAddress(walletAddress)
                .map(ArtistIdentity::isVerified)
                .orElse(false);
    }
}

File: app02_cpp_nft/backend/src/identity/MockKycVerificationService.cpp
#include "identity/MockKycVerificationService.h"

#include "util/Time.h"
#include "util/Uuid.h"

namespace nftcerts::identity {

ArtistIdentity MockKycVerificationService::verify(const std::string& walletAddress, const std::string& did,
                                                    const std::string& email) {
    ArtistIdentity identity;
    auto existing = repository_.findByWalletAddress(walletAddress);
    if (existing.has_value()) {
        identity = *existing;
    } else {
        identity.id = util::randomUuid();
        identity.walletAddress = walletAddress;
    }

    identity.did = did;
    identity.email = email;
    identity.verified = true;
    identity.verifiedAt = util::nowIso8601();

    return repository_.save(identity);
}

bool MockKycVerificationService::isVerified(const std::string& walletAddress) {
    auto identity = repository_.findByWalletAddress(walletAddress);
    return identity.has_value() && identity->verified;
}

}  // namespace nftcerts::identity
